File: js/test/regression-165.js
/**
 * Regression tests for issue #165.
 *
 * A runtime that aborts on its own heap limit (`FATAL ERROR: Reached heap limit
 * ...`) dies below the container limit, so every container signal `start`
 * consults is correct and useless: `State.OOMKilled` is `false` and the cgroup
 * `oom_kill` counter is `0`. The only evidence is what the runtime printed into
 * the log, which is why `--status` now carries `memoryExhausted` /
 * `memoryExhaustedReason` alongside `oomKilled`, and why the kept-container
 * footer no longer asserts a bare `oomKilled=false` next to a fatal marker.
 */

const { describe, it, expect, beforeEach, afterEach } = require('bun:test');
const { spawnSync } = require('child_process');
const fs = require('fs');
const os = require('os');
const path = require('path');

const {
  ExecutionRecord,
  ExecutionStore,
} = require('../src/lib/execution-store');
const {
  buildAttachedDockerKeptMessage,
  buildDetachedDockerCompletionScript,
  buildDockerKeptReasonSnippet,
} = require('../src/lib/docker-cleanup');

const TEST_APP_FOLDER = path.join(
  os.tmpdir(),
  `regression-165-${process.pid}-${Date.now()}`
);
const CLI_PATH = path.join(__dirname, '../src/bin/cli.js');

const HEAP_LIMIT_MARKER =
  'FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory';

function cleanupTestDir() {
  fs.rmSync(TEST_APP_FOLDER, { recursive: true, force: true });
}

function runCli(args) {
  const result = spawnSync(process.execPath, [CLI_PATH, ...args], {
    encoding: 'utf8',
    env: { ...process.env, START_APP_FOLDER: TEST_APP_FOLDER },
    timeout: 20000,
  });
  return {
    stdout: result.stdout || '',
    stderr: result.stderr || '',
    exitCode: result.status,
  };
}

/**
 * Write a log that ends the way the incident log does: command output, the
 * dying runtime's fatal marker, an optional native stack trace, and the
 * terminal footer `start` appends itself.
 */
function writeOomLog(
  name,
  { stackTraceBytes = 0, marker = HEAP_LIMIT_MARKER }
) {
  const logPath = path.join(TEST_APP_FOLDER, name);
  const stackFrame =
    ' 1: 0x757658 node::OOMErrorHandler(char const*, v8::OOMDetails const&) [node]\n';
  const frames = stackTraceBytes
    ? stackFrame.repeat(Math.ceil(stackTraceBytes / stackFrame.length))
    : '';
  fs.writeFileSync(
    logPath,
    [
      'building...',
      '',
      '<--- Last few GCs --->',
      '[1:0xa485000]  790 ms: Mark-Compact 66.6 (99.8) -> 64.4 (194.2) MB',
      '',
      marker,
      '----- Native stack trace -----',
      frames,
      'Docker container "docker-165" exited with code 139',
      'Container kept because the command failed.',
      '',
      '==================================================',
      'Finished: 2026-09-03 08:14:20.707',
      'Exit Code: 139',
      '',
    ].join('\n'),
    'utf8'
  );
  return logPath;
}

function saveRecord(store, overrides = {}) {
  const record = new ExecutionRecord({
    command: overrides.command || 'node build.mjs',
    pid: 4242,
    logPath: overrides.logPath || null,
    workingDirectory: '/tmp',
    shell: '/bin/bash',
    options: overrides.options || {},
  });
  if (overrides.exitCode !== undefined) {
    record.complete(overrides.exitCode);
  }
  store.save(record);
  return record;
}

function statusJson(uuid) {
  const result = runCli(['--status', uuid, '--output-format', 'json']);
  expect(result.stderr).toBe('');
  expect(result.exitCode).toBe(0);
  return JSON.parse(result.stdout);
}

describe('Issue #165: memory exhaustion observed from the log', () => {
  let store;

  beforeEach(() => {
    cleanupTestDir();
    fs.mkdirSync(TEST_APP_FOLDER, { recursive: true });
    store = new ExecutionStore({
      appFolder: TEST_APP_FOLDER,
      useLinks: false,
    });
  });

  afterEach(() => {
    cleanupTestDir();
  });

  it('reports memoryExhausted for an attached docker session that self-aborted', () => {
    const record = saveRecord(store, {
      logPath: writeOomLog('attached.log', {}),
      exitCode: 139,
      options: { isolated: 'docker', isolationMode: 'attached' },
    });

    const parsed = statusJson(record.uuid);

    expect(parsed.exitCode).toBe(139);
    // The container flag is still absent/false — the observation does not
    // contradict it, it complements it.
    expect(parsed.oomKilled).toBeUndefined();
    expect(parsed.memoryExhausted).toBe(true);
    expect(parsed.memoryExhaustedReason).toBe(HEAP_LIMIT_MARKER);
    expect(parsed.exitReason).toBe('memory-exhaustion (v8-heap-limit)');
  });

  it('finds a marker pushed far from EOF by the native stack trace', () => {
    const logPath = writeOomLog('long-trace.log', {
      stackTraceBytes: 40 * 1024,
    });
    expect(fs.statSync(logPath).size).toBeGreaterThan(32 * 1024);
    const record = saveRecord(store, { logPath, exitCode: 134 });

    const parsed = statusJson(record.uuid);

    expect(parsed.memoryExhausted).toBe(true);
    expect(parsed.memoryExhaustedReason).toBe(HEAP_LIMIT_MARKER);
  });

  it('detects a Rust allocation failure', () => {
    const record = saveRecord(store, {
      logPath: writeOomLog('rust.log', {
        marker: 'memory allocation of 1073741824 bytes failed',
      }),
      exitCode: 134,
    });

    const parsed = statusJson(record.uuid);

    expect(parsed.memoryExhausted).toBe(true);
    expect(parsed.memoryExhaustedReason).toBe(
      'memory allocation of 1073741824 bytes failed'
    );
  });

  it('never turns a clean run into a reported memory failure', () => {
    // The marker is present in the output — the command merely printed it —
    // but the run succeeded, so there is nothing to explain (issue #151's rule:
    // an observation, never a verdict).
    const logPath = path.join(TEST_APP_FOLDER, 'quoted.log');
    fs.writeFileSync(
      logPath,
      [
        `grep found: ${HEAP_LIMIT_MARKER}`,
        '==================================================',
        'Finished: 2026-09-03 08:14:20.707',
        'Exit Code: 0',
        '',
      ].join('\n'),
      'utf8'
    );
    const record = saveRecord(store, { logPath, exitCode: 0 });

    const parsed = statusJson(record.uuid);

    expect(parsed.exitCode).toBe(0);
    expect(parsed.memoryExhausted).toBeUndefined();
    expect(parsed.memoryExhaustedReason).toBeUndefined();
  });

  it('leaves an ordinary failure without a memory observation', () => {
    const logPath = path.join(TEST_APP_FOLDER, 'plain-failure.log');
    fs.writeFileSync(logPath, 'error: missing file\n', 'utf8');
    const record = saveRecord(store, { logPath, exitCode: 1 });

    const parsed = statusJson(record.uuid);

    expect(parsed.memoryExhausted).toBeUndefined();
  });

  it('shows the observation in the human-readable and links-notation output', () => {
    const record = saveRecord(store, {
      logPath: writeOomLog('formats.log', {}),
      exitCode: 139,
    });

    const text = runCli(['--status', record.uuid, '--output-format', 'text']);
    expect(text.exitCode).toBe(0);
    expect(text.stdout).toContain('Memory Exhausted:  true');
    expect(text.stdout).toContain(`Memory Evidence:   ${HEAP_LIMIT_MARKER}`);

    const lino = runCli(['--status', record.uuid]);
    expect(lino.exitCode).toBe(0);
    expect(lino.stdout).toContain('memoryExhausted true');
    expect(lino.stdout).toContain('memoryExhaustedReason');
  });
});

describe('Issue #165: kept-container footer', () => {
  const evaluateReason = (exitCode, oomKilled) => {
    const snippet = buildDockerKeptReasonSnippet();
    const result = spawnSync(
      'sh',
      [
        '-c',
        `__start_command_exit=${exitCode}; __start_command_oom=${oomKilled}; ${snippet}; printf '%s' "$__start_command_reason"`,
      ],
      { encoding: 'utf8' }
    );
    expect(result.status).toBe(0);
    return result.stdout;
  };

  it('does not assert a bare oomKilled=false for a self-abort exit code', () => {
    for (const exitCode of [134, 139]) {
      const reason = evaluateReason(exitCode, 'false');
      expect(reason).toContain(`exitCode=${exitCode} oomKilled=false`);
      expect(reason).toContain('invisible to this flag');
    }
  });

  it('keeps the plain reason for other exit codes and for a real OOM kill', () => {
    expect(evaluateReason(1, 'false')).toBe('exitCode=1 oomKilled=false');
    expect(evaluateReason(137, 'true')).toBe('exitCode=137 oomKilled=true');
  });

  it('embeds the reason computation in the detached completion script', () => {
    const script = buildDetachedDockerCompletionScript(
      'container-165',
      'default',
      '/tmp/regression-165.log'
    );
    expect(script).toContain('__start_command_reason=');
    expect(script).toContain('Reason: %s');
  });
});

describe('Issue #165: attached kept-container message', () => {
  beforeEach(() => {
    cleanupTestDir();
    fs.mkdirSync(TEST_APP_FOLDER, { recursive: true });
  });

  afterEach(() => {
    cleanupTestDir();
  });

  it('names the evidence the runtime printed into the log', () => {
    const message = buildAttachedDockerKeptMessage({
      containerName: 'container-165',
      exitCode: 139,
      oomKilled: false,
      logPath: writeOomLog('attached-message.log', {}),
    });

    expect(message).toContain('Container kept because the command failed.');
    expect(message).toContain(
      `Memory exhaustion detected in the log: ${HEAP_LIMIT_MARKER}`
    );
    expect(message).toContain('Remove when done: docker rm -f container-165');
  });

  it('stays silent for an ordinary failure', () => {
    const logPath = path.join(TEST_APP_FOLDER, 'attached-plain.log');
    fs.writeFileSync(logPath, 'error: missing file\n', 'utf8');

    const message = buildAttachedDockerKeptMessage({
      containerName: 'container-165',
      exitCode: 1,
      oomKilled: false,
      logPath,
    });

    expect(message).not.toContain('Memory exhaustion');
  });
});

File: rust/tests/regression_165.rs
//! Regression tests for issue #165:
//! "Memory exhaustion reported by the runtime itself is invisible to `--status`"
//!
//! A runtime that aborts on its own heap limit (`FATAL ERROR: Reached heap
//! limit ...`) dies below the container limit, so every container signal
//! `start` consults is correct and useless: `State.OOMKilled` is `false` and
//! the cgroup `oom_kill` counter is `0`. The only evidence is what the runtime
//! printed into the log, which is why `--status` now carries `memoryExhausted`
//! / `memoryExhaustedReason` alongside `oomKilled`, and why the kept-container
//! footer no longer asserts a bare `oomKilled=false` next to a fatal marker.
//!
//! Reference: https://github.com/link-foundation/start/issues/165

use serde_json::Value;
use start_command::execution_store::{
    ExecutionRecord, ExecutionRecordOptions, ExecutionStore, ExecutionStoreOptions,
};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempfile::TempDir;

const SEPARATOR: &str = "==================================================";
const HEAP_LIMIT_MARKER: &str =
    "FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory";

fn create_test_store(app_folder: &Path) -> ExecutionStore {
    ExecutionStore::with_options(ExecutionStoreOptions {
        app_folder: Some(app_folder.to_path_buf()),
        use_links: Some(false),
        verbose: false,
    })
}

fn run_cli(app_folder: &Path, args: &[&str]) -> (String, String, i32) {
    let output = Command::new(env!("CARGO_BIN_EXE_start"))
        .args(args)
        .env("START_APP_FOLDER", app_folder)
        .env("START_DISABLE_AUTO_ISSUE", "1")
        .env("START_DISABLE_LOG_UPLOAD", "1")
        .output()
        .expect("failed to run the start binary");
    (
        String::from_utf8_lossy(&output.stdout).to_string(),
        String::from_utf8_lossy(&output.stderr).to_string(),
        output.status.code().unwrap_or(-1),
    )
}

/// Write a log that ends the way the incident log does: command output, the
/// dying runtime's fatal marker, an optional native stack trace, and the
/// terminal footer `start` appends itself.
fn write_oom_log(dir: &Path, name: &str, marker: &str, stack_trace_bytes: usize) -> PathBuf {
    let log_path = dir.join(name);
    let frame = " 1: 0x757658 node::OOMErrorHandler(char const*, v8::OOMDetails const&) [node]\n";
    let frames = frame.repeat(stack_trace_bytes.div_ceil(frame.len()));
    fs::write(
        &log_path,
        format!(
            "building...\n\n<--- Last few GCs --->\n{marker}\n----- Native stack trace -----\n\
             {frames}Docker container \"docker-165\" exited with code 139\n\
             Container kept because the command failed.\n\n{SEPARATOR}\n\
             Finished: 2026-09-03 08:14:20.707\nExit Code: 139\n"
        ),
    )
    .unwrap();
    log_path
}

fn save_record(store: &ExecutionStore, log_path: &Path, exit_code: i32) -> ExecutionRecord {
    let mut record = ExecutionRecord::with_options(ExecutionRecordOptions {
        command: "node build.mjs".to_string(),
        pid: Some(4242),
        log_path: Some(log_path.to_string_lossy().to_string()),
        working_directory: Some("/tmp".to_string()),
        shell: Some("/bin/bash".to_string()),
        ..Default::default()
    });
    record.complete(exit_code);
    store.save(&record).unwrap();
    record
}

fn status_json(app_folder: &Path, uuid: &str) -> Value {
    let (stdout, stderr, exit_code) =
        run_cli(app_folder, &["--status", uuid, "--output-format", "json"]);
    assert_eq!(exit_code, 0, "stderr: {stderr}");
    serde_json::from_str(&stdout).unwrap()
}

#[test]
fn test_status_reports_memory_exhaustion_for_a_self_aborted_run() {
    let temp = TempDir::new().unwrap();
    let store = create_test_store(temp.path());
    let log_path = write_oom_log(temp.path(), "attached.log", HEAP_LIMIT_MARKER, 0);
    let record = save_record(&store, &log_path, 139);

    let parsed = status_json(temp.path(), &record.uuid);

    assert_eq!(parsed["exitCode"], 139);
    // The container flag is still absent/false - the observation does not
    // contradict it, it complements it.
    assert!(parsed.get("oomKilled").is_none() || parsed["oomKilled"].is_null());
    assert_eq!(parsed["memoryExhausted"], true);
    assert_eq!(parsed["memoryExhaustedReason"], HEAP_LIMIT_MARKER);
    assert_eq!(parsed["exitReason"], "memory-exhaustion (v8-heap-limit)");
}

#[test]
fn test_status_finds_a_marker_pushed_far_from_the_end_of_the_log() {
    let temp = TempDir::new().unwrap();
    let store = create_test_store(temp.path());
    let log_path = write_oom_log(temp.path(), "long-trace.log", HEAP_LIMIT_MARKER, 40 * 1024);
    assert!(fs::metadata(&log_path).unwrap().len() > 32 * 1024);
    let record = save_record(&store, &log_path, 134);

    let parsed = status_json(temp.path(), &record.uuid);

    assert_eq!(parsed["memoryExhausted"], true);
    assert_eq!(parsed["memoryExhaustedReason"], HEAP_LIMIT_MARKER);
}

#[test]
fn test_status_detects_a_rust_allocation_failure() {
    let temp = TempDir::new().unwrap();
    let store = create_test_store(temp.path());
    let marker = "memory allocation of 1073741824 bytes failed";
    let log_path = write_oom_log(temp.path(), "rust.log", marker, 0);
    let record = save_record(&store, &log_path, 134);

    let parsed = status_json(temp.path(), &record.uuid);

    assert_eq!(parsed["memoryExhausted"], true);
    assert_eq!(parsed["memoryExhaustedReason"], marker);
}

#[test]
fn test_status_never_turns_a_clean_run_into_a_memory_failure() {
    // The marker is present in the output - the command merely printed it -
    // but the run succeeded, so there is nothing to explain (issue #151's
    // rule: an observation, never a verdict).
    let temp = TempDir::new().unwrap();
    let store = create_test_store(temp.path());
    let log_path = temp.path().join("quoted.log");
    fs::write(
        &log_path,
        format!(
            "grep found: {HEAP_LIMIT_MARKER}\n{SEPARATOR}\n\
             Finished: 2026-09-03 08:14:20.707\nExit Code: 0\n"
        ),
    )
    .unwrap();
    let record = save_record(&store, &log_path, 0);

    let parsed = status_json(temp.path(), &record.uuid);

    assert_eq!(parsed["exitCode"], 0);
    assert!(parsed.get("memoryExhausted").is_none() || parsed["memoryExhausted"].is_null());
}

#[test]
fn test_status_leaves_an_ordinary_failure_without_a_memory_observation() {
    let temp = TempDir::new().unwrap();
    let store = create_test_store(temp.path());
    let log_path = temp.path().join("plain-failure.log");
    fs::write(&log_path, "error: missing file\n").unwrap();
    let record = save_record(&store, &log_path, 1);

    let parsed = status_json(temp.path(), &record.uuid);

    assert!(parsed.get("memoryExhausted").is_none() || parsed["memoryExhausted"].is_null());
}

#[test]
fn test_status_shows_the_observation_in_the_human_readable_output() {
    let temp = TempDir::new().unwrap();
    let store = create_test_store(temp.path());
    let log_path = write_oom_log(temp.path(), "formats.log", HEAP_LIMIT_MARKER, 0);
    let record = save_record(&store, &log_path, 139);

    let (stdout, stderr, exit_code) = run_cli(
        temp.path(),
        &["--status", &record.uuid, "--output-format", "text"],
    );

    assert_eq!(exit_code, 0, "stderr: {stderr}");
    assert!(
        stdout.contains("Memory Exhausted:  true"),
        "stdout: {stdout}"
    );
    assert!(
        stdout.contains(&format!("Memory Evidence:   {HEAP_LIMIT_MARKER}")),
        "stdout: {stdout}"
    );
}
